movement-test/src/main.c

```diff
@@ -64,80 +78,306 @@ UINT8 update_sprite_animation(UINT8 sprite_id, UINT8 *anim, UINT8 direction, UIN
+      for (UINT8 j = 0; j < 3; j++) {
+        INT8 block_x = (effective_x + i - 1);
+        INT8 block_y = (effective_y + j - 1);
+
+        if (block_x < 0 || block_x >= (INT8) ROOM_WIDTH + 2 || block_y < 0 || block_y >= (INT8) ROOM_HEIGHT + 2)  // out of bounds
+          continue;
+
+        INT16 k = block_y * (ROOM_WIDTH + 2) + block_x;
+
+        if (TILEMAP[k] != 0) {
+          block.pos.x = block_x << 3;
+          block.pos.y = block_y << 3;
+
+          if (rect_rect_collision(&new_player, &block)) {
+            VEC_DIFF diff = {0, 0};
+            rect_rect_penetration(&(player.pos), &(new_player.pos), &(player.size), &block, &diff);
+            new_player.pos.y += diff.dy;
+            total_diff.dy = 0;
+          }
+        }
+      }
+    }
+
+    player.pos.x = new_player.pos.x;
+    player.pos.y = new_player.pos.y;
+
+  }
+}
+
+void init_dash_state() {
+    is_dashing = 0;
+    compteur_dash = 0;
+}
+
+
+void main(void) {
+    init_player_state();
+    init_dash_state();
+
+    init_graphics();
+
     while (1) {
         // Wait for v-blank (screen refresh)
         wait_vbl_done();
 
-        dx = 0;
-        dy = 0;
-
-        // Read joypad keys to know if the player is walking
-        // and in which direction
-        keys = joypad();
-        if (keys & J_UP) {
-            player_direction = PLAYER_DIRECTION_UP;
-            dy -= 1;
-        } if (keys & J_DOWN) {
-            player_direction = PLAYER_DIRECTION_DOWN;
-            dy += 1;
-        } if (keys & J_LEFT) {
-            player_direction = PLAYER_DIRECTION_LEFT;
-            dx -= 1;
-        } if (keys & J_RIGHT) {
-            player_direction = PLAYER_DIRECTION_RIGHT;
-            dx += 1;
-        }
+        read_input();
 
-       is_player_walking = (dx != 0 || dy != 0);
+        handle_dash();
 
-        // Update the player position if it is walking
-        if (is_player_walking) {
+        handle_collisions();
 
-            player_x += dx;
-            player_y += dy;
 
-            move_sprite(PLAYER_SPRITE_ID, player_x, player_y);
+        // Do NOT move this near update_sprite_animation...
+        move_sprite(PLAYER_SPRITE_ID, player.pos.x + SPRITE_OFFSET_X + scroll_x, player.pos.y + SPRITE_OFFSET_Y + scroll_y);
 
-            // We do not update the animation on each frame: the animation
-            // will be too quick. So we skip frames
-            frame_skip -= 1;
-            if (!frame_skip) {
-                frame_skip = 8;
-            } else {
-                continue;
-            }
+        // We do not update the animation on each frame: the animation
+        // will be too quick. So we skip frames
+        frame_skip -= 1;
+        if (!frame_skip) {
+          frame_skip = 8;
         } else {
+          continue;
+        }
+
+        // XXX: this is not necessarily the right condition, but oh well.
+        if (dx == 0 && dy == 0){
             player_animation_frame = 0;
         }
```

Comment:
I'm still not sure where we should put this code, it's needed for the player animation but this is all quite fragile. Should probably be addressed later on.

Reply:
I'm not sure either. For now it works. It might break when we'll add the dash animations